libs/aws/langchain_aws/agents/utils.py

```diff
@@ -0,0 +1,287 @@
+) -> Any:
+    """
+    Construct the boto3 session
+    """
+    if credentials_profile_name:
+        session = boto3.Session(profile_name=credentials_profile_name)
+    else:
+        # use default credentials
+        session = boto3.Session()
+    client_params = {
+        "config": Config(
+            connect_timeout=120, read_timeout=120, retries={"max_attempts": 0}
+        )
+    }
+    if region_name:
+        client_params["region_name"] = region_name
+    if endpoint_url:
+        client_params["endpoint_url"] = endpoint_url
+    return client_params, session
+
+
+def parse_agent_response(response: Any) -> OutputType:
+    """
+    Parses the raw response from Bedrock Agent
+
+    Args:
+        response: The raw response from Bedrock Agent
+
+    Returns
+        Either a BedrockAgentAction or a BedrockAgentFinish
+    """
+    response_text = ""
+    event_stream = response["completion"]
+    session_id = response["sessionId"]
+    trace_log_elements = []
+    files = []
+    for event in event_stream:
+        if "trace" in event:
+            trace_log_elements.append(event["trace"])
+
+        if "returnControl" in event:
+            response_text = json.dumps(event)
+            break
+
+        if "chunk" in event:
+            response_text = event["chunk"]["bytes"].decode("utf-8")
+            
+        if "files" in event:
+            files = event["files"]["files"]
+
+    trace_log = json.dumps(trace_log_elements)
+
+    agent_finish = BedrockAgentFinish(
+        return_values={"output": response_text, "files": files},
+        log=response_text,
+        session_id=session_id,
+        trace_log=trace_log,
+    )
+    if not response_text:
+        return agent_finish
+
+    if "returnControl" not in response_text:
+        return agent_finish
+
+    return_control = json.loads(response_text).get("returnControl")
+    if not return_control:
+        return agent_finish
+
+    invocation_inputs = return_control.get("invocationInputs")
+    if not invocation_inputs:
+        return agent_finish
+
+    try:
+        invocation_input = invocation_inputs[0].get("functionInvocationInput", {})
+        action_group = invocation_input.get("actionGroup", "")
+        function = invocation_input.get("function", "")
+        parameters = invocation_input.get("parameters", [])
+        parameters_json = {}
+        for parameter in parameters:
+            parameters_json[parameter.get("name")] = parameter.get("value", "")
+
+        tool = f"{action_group}::{function}"
+        if _DEFAULT_ACTION_GROUP_NAME in action_group:
+            tool = f"{function}"
+        return [
+            BedrockAgentAction(
+                tool=tool,
+                tool_input=parameters_json,
+                log=response_text,
+                session_id=session_id,
+                trace_log=trace_log,
+            )
+        ]
+    except Exception as ex:
+        raise Exception("Parse exception encountered {}".format(repr(ex)))
```

Comment:
Is there an exception hierarchy that we can add onto here? `Exception` is broad.

Reply:
Unsure of this one, keeping it broad to encompass all validation + serviceUnavailable exceptions.

Reply:
Agree with @nateglims comment here, catching specific exceptions and then providing a simplified error message for the user will be better, but don't find it blocking for this PR.

nit: Can you use the form `raise Exception(...) from ex`, so the original exception is chained.

Reply:
Added some more specific exceptions and chaining.